independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java

```diff
@@ -0,0 +1,245 @@
+    }
+
+    public Set<String> getLocalClassResourceNames() {
+        return localClasses;
+    }
+
+    public boolean isParentFirst(String resource) {
+        return parentFirstResources.contains(resource);
+    }
+
+    public boolean isBanned(String resource) {
+        return bannedResources.contains(resource);
+    }
+
+    // it's tempting to use an Optional here but let's avoid the additional allocation
+    public ClassPathElement getFirstClassPathElement(String resource) {
+        ClassPathElement transformedClassClassPathElement = transformedClasses.get(resource);
+        if (transformedClassClassPathElement != null) {
+            return transformedClassClassPathElement;
+        }
+
+        ClassPathElement[] candidates = resourceMapping.get(getResourceKey(resource));
+        if (candidates == null) {
+            return null;
+        }
+
+        for (ClassPathElement candidate : candidates) {
+            if (candidate.getProvidedResources().contains(resource)) {
+                return candidate;
+            }
+        }
+
+        return null;
+    }
+
+    public List<ClassPathElement> getClassPathElements(String resource) {
+        ClassPathElement transformedClassClassPathElement = transformedClasses.get(resource);
+        if (transformedClassClassPathElement != null) {
+            return List.of(transformedClassClassPathElement);
+        }
+
+        ClassPathElement[] candidates = resourceMapping.get(getResourceKey(resource));
+        if (candidates == null) {
+            return List.of();
+        }
+
+        if (candidates.length == 1) {
+            if (candidates[0].getProvidedResources().contains(resource)) {
+                return List.of(candidates[0]);
+            }
+
+            return List.of();
+        }
+
+        List<ClassPathElement> classPathElements = new ArrayList<>(candidates.length);
+        for (ClassPathElement candidate : candidates) {
+            if (candidate.getProvidedResources().contains(resource)) {
+                classPathElements.add(candidate);
+            }
+        }
+        return Collections.unmodifiableList(classPathElements);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Returns a key that tries to find a good compromise between reducing the size of the index and providing good
+     * performances.
+     * <p>
+     * Probably something we will have to tweak for corner cases but let's try to keep it fast.
+     */
+    private static String getResourceKey(String resource) {
+        // we don't really care about this part, it can be slower
+        if (resource.startsWith(META_INF_MAVEN)) {
+            return META_INF_MAVEN;
+        }
+        if (resource.startsWith(META_INF_SERVICES)) {
+            // for services, we want to reference the full path
+            return resource;
+        }
+        StringBuilder prefix = new StringBuilder();
```

Comment:
Any chance we can make an okay guess of the length?

Reply:
Not really. And interestingly enough, I made some JMH benchmarks this week-end for SmallRye Config (so totally unrelated) and ended up with the version specifying the length being slightly slower.

Note that I conducted some JMH testing of this method to make sure it's fast (not saying it can't be improved but it's not a naive approach to it - see follow-up commit).